querqy-core/src/main/java/querqy/rewrite/contrib/NumberQueryRewriter.java

```diff
@@ -0,0 +1,150 @@
+    int numberOfClauses = 0;
+    int i = 1;
+
+    final boolean acceptGeneratedTerms;
+
+    public NumberQueryRewriter(){
+        this(false);
+    }
+
+    public NumberQueryRewriter(final boolean acceptGeneratedTerms) {
+        this.acceptGeneratedTerms = acceptGeneratedTerms;
+    }
+
+    @Override
+    public RewriterOutput rewrite(final ExpandedQuery query, final SearchEngineRequestAdapter requestAdapter) {
+        final QuerqyQuery<?> userQuery = query.getUserQuery();
+        if (userQuery != null && userQuery instanceof Query){
+            previousTerm = null;
+            termsToAdd = new LinkedList<>();
+            numberOfClauses = ((Query) userQuery).getClauses().size();
+            i = 1;
+            visit((Query) userQuery);
+            for (Term term : termsToAdd) {
+                term.getParent().addClause(term);
+            }
+        }
+        return RewriterOutput.builder().expandedQuery(query).build();
+    }
+
+    @Override
+    public Node visit(final DisjunctionMaxQuery dmq) {
+
+        final List<DisjunctionMaxClause> clauses = dmq.getClauses();
+
+        if (clauses != null) {
+
+            switch (clauses.size()) {
+
+                case 0: break;
+
+                case 1: super.visit(dmq); break;
+
+                default:
+
+                    if (acceptGeneratedTerms) {
+
+                        throw new IllegalArgumentException("cannot handle more than one DMQ clause");
+
+                    } else {
+
+                        DisjunctionMaxClause nonGeneratedClause = null;
+
+                        for (DisjunctionMaxClause clause: clauses) {
+
+                            if (!clause.isGenerated()) {
+                                // second non-generated clause - cannot handle this
+                                if (nonGeneratedClause != null) {
+                                    throw new IllegalArgumentException("cannot handle more than one non-generated DMQ clause");
+                                }
+                                nonGeneratedClause = clause;
+                            }
+                        }
+
+                        if (nonGeneratedClause != null) {
+                            nonGeneratedClause.accept(this);
+                        }
+                    }
+
+            }
+
+
+        }
+
+        return null;
+
+    }
+
+    @Override
+    public Node visit(final Term term) {
+        if (previousTerm != null
+                && eq(previousTerm.getField(), term.getField())
+                && (term.isGenerated() == acceptGeneratedTerms || !term.isGenerated())
+                && (previousTerm.isGenerated() == acceptGeneratedTerms || !previousTerm.isGenerated())) {
+            //if previousTerm and term are digits-only add them to a list
+            if (previousTerm.toString().matches("\\d+") && term.toString().matches("\\d+") && previousTerm.toString().concat(term.toString()).length() > 2) {
```

Comment:
This might be inefficient (and we want to be as fast as possible here because all query terms have to pass this for all queries): we're calling `toString()`, compile a regex and then do the matching for both terms each time, and then we call `toString()` for both terms again, concatenate them and then check the length.

Ideas for optimisation based on the fact that `Term` implements `java.lang.CharSequence`:

```
final boolean isDigit(final Term term):
   if (term.length() < 1) return false; // not sure if we need this
   for (int pos = 0, size=term.length(); pos < size; pos++) {
      if (!Character.isDigit(term.charAt(pos))) {
          return false;
      }
    }
    return true;
}
```
... and replace `previousTerm.toString().matches("\\d+")`  with `isDigit(previousTerm)`  etc.

`previousTerm.toString().concat(term.toString()).length() > 2)` could probably become `(previousTerm.length() + term.length()) > 2`. 

On this note: Is there a min length of 3 of the combined numeric string implied here?  - That number should then be made configurable.



   }